### ISSUE TITLE
Question about running batch inference with PyTorch Tensor as input

### ISSUE THREAD

**Issue by Renzzauw** [ultralytics/yolov5#6726]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi!

### Problem introduction
I am trying to run batch inference on batches of video frames obtained from NVIDIA DALI, but I am having some issues with implementing this. I am trying to make this work, as performance is of high importance, which I try to maximize as much as possible.

### What I tried so far
I simply convert and reshape the output of DALI to a PyTorch tensor of shape (batch_size, 3, height, width) and then feed this to my yolov5 model, which I load using the PyTorch hub, as shown in the batch inference [example code](https://github.com/ultralytics/yolov5/issues/36). This resulted in various errors with regard to shapes not matching, so I assume my input is not of the correct shape or not properly pre-processed.

I did some digging in the yolov5 code to see how exactly batch inference works and what [types of input ](https://github.com/ultralytics/yolov5/blob/c43f13557185403631d7eae804413ccf27ae9a2a/models/common.py#L515) are supported. I found that when you input a list of images (e.g. of type ndarray or PIL), these images are [pre-processed](https://github.com/ultralytics/yolov5/blob/c43f13557185403631d7eae804413ccf27ae9a2a/models/common.py#L526), so they are made suitable for running inference. However, when I input a PyTorch Tensor, I see this is directly fed [to the model](https://github.com/ultralytics/yolov5/blob/c43f13557185403631d7eae804413ccf27ae9a2a/models/common.py#L521), seemingly without any pre-processing (as far as I can understand the code). 

### Question
My question is, what are the required steps to make a PyTorch Tensor work as input for batch inference? The forward function arguments comments suggest the width or height should be equal to the size parameter and the values should be within range [0, 1], but this does not seem to be all the required steps. Is there any functions available within this repo to pre-process the Tensor?

### Alternative I tried
I did some alternative tests with passing a list of video frames (ndarrays of shape (height, width, 3)) instead to the model, but the performance gain seemed to be limited compared to simply running on batch size 1. Hence, I am trying to make batch inference work.

I did not seem to find any resources online or in existing issues that answer my very specific question. Thanks in advance for helping me out with this question. If I need to provide any more details, feel free to tell me so!

**Comment by glenn-jocher** [comment]
@Renzzauw 👋 Hello! Thanks for asking about **inference speed issues**. For our official YOLOv5 batch-size speed study see https://community.ultralytics.com/t/yolov5-study-batch-size-vs-speed/31

YOLOv5 🚀 can be run on CPU (i.e. `--device cpu`, slow) or GPU if available (i.e. `--device 0`, faster). You can determine your inference device by viewing the YOLOv5 console output:

### detect.py inference
```bash
python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images/
```
<img width="944" alt="detect.py" src="https://user-images.githubusercontent.com/26833433/140100842-e130b5f7-64cb-46c9-9386-72abadd19828.png">

### YOLOv5 PyTorch [Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) inference
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()  # or .show(), .save()
# Speed: 631.5ms pre-process, 19.2ms inference, 1.6ms NMS per image at shape (2, 3, 640, 640)
```

### Increase Speeds

If you would like to increase your inference speed some options are:

- Use batched inference with [YOLOv5 PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
- Reduce `--img-size`, i.e. 1280 -> 640 -> 320
- Reduce model size, i.e. YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s -> YOLOv5n
- Use half precision FP16 inference with `python detect.py --half` and `python val.py --half`
- Use a faster GPUs, i.e.: P100 -> V100 -> A100
- [Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) to ONNX for up to 3x CPU speedup
- [Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) to TensorRT for up to 5x GPU speedup
- Use a free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Good luck 🍀 and let us know if you have any other questions!

**Comment by Renzzauw** [comment]
Hi @glenn-jocher and thank you for your reply.

Thank you for providing some resources. I have studied most of these already, and I'm afraid that they do not really provide answer to my question. I have taken various steps you mentioned in the **Increase Speeds** part you linked, yet I'm specifically interested in more information about batched inference. However, the examples you've linked, which I studied a lot recently, simply do not cover any details in regard to inputting a PyTorch Tensor.

**Comment by glenn-jocher** [comment]
@Renzzauw YOLOv5 PyTorch Hub models support torch tensor inputs at any batch size, however since we can not determine the preprocessing steps used (i.e. letterboxing, padding etc.) we can not implement a postprocessing strategy. Therefore torch inference types act as pass-throughs, enabling you to create your own pre and post processing workflows.

**Comment by glenn-jocher** [comment]
See AutoShape forward method for details:

https://github.com/ultralytics/yolov5/blob/c43f13557185403631d7eae804413ccf27ae9a2a/models/common.py#L507-L524

**Comment by Renzzauw** [comment]
@glenn-jocher Thank you for your reply, this makes it clear how to use PyTorch tensors. I will look into implementing my own pre- and post-processing steps.
